### DISCUSSION TITLE
Can't push, files too big

### DISCUSSION THREAD

**Discussion by WillSamuel18** [community/community#113613] [Other Feature Feedback, Questions, & Ideas]
### Select Topic Area

Question

### Body

Hi, I am working in R Studio and using github to store my code and data. I am running into issues on one project since my files are too large to push (>100 MB). I have since deleted the files from my local folder, but GitHub is still trying to push them from the previous commits. This makes it so that I am unable to push to the remote repo. Any suggestions? Is there a way I can bypass the commit with those images and jump straight to the current project which doesn't include those large files? I know I could update these manually through the website, but I am trying to reconnect it through R Studio which is preferable for me. 
Here is the error message. These two files have since been deleted:

_>>> C:/Program Files/Git/cmd/git.exe push origin HEAD:refs/heads/main
remote: error: Trace: 5531ef38053c05a217b0cd5bcc42c9408b016659a30a904526ef682580cd142e        
remote: error: See https://gh.io/lfs for more information.        
remote: error: File Figure 3 option2.tiff is 314.71 MB; this exceeds GitHub's file size limit of 100.00 MB        
remote: error: File Figure 3.tiff is 314.71 MB; this exceeds GitHub's file size limit of 100.00 MB        
remote: error: GH001: Large files detected. You may want to try Git Large File Storage - https://git-lfs.github.com.        
To https://github.com/WillSamuel18/Arctic-Grayling-Morphometrics-Public
 ! [remote rejected] HEAD -> main (pre-receive hook declined)
error: failed to push some refs to 'https://github.com/WillSamuel18/Arctic-Grayling-Morphometrics-Public'_

**Comment by JasonLuuk** [comment]
1. Backup your repository before making any irreversible changes.
2. Use the git filter-branch command to remove files from all commits:
```
git filter-branch --force --index-filter "git rm --cached --ignore-unmatch <path_to_large_file>" --prune-empty --tag-name-filter cat -- --all
```
3. Replace <path_to_large_file> with the path to the file you want to remove from your repository's history
4. Force push the changes to your GitHub repository:
```
git push origin --force --all
```
***If you want to commit a large file, try [Git Large File Storage](https://git-lfs.com/)***

**Reply by WillSamuel18** [reply]
Thanks for your help! I don't think that worked, can you see what the issue is here? 

C:\Users\npwil\OneDrive\Desktop\School\Grad School\Side Quests\Grayling morphometrics project\Arctic-Grayling-Morphometrics-Public>git filter-branch --force --index-filter "git rm --cached --ignore-unmatch <C:\Users\npwil\OneDrive\Desktop\School\Grad School\Side Quests\Grayling morphometrics project\Arctic-Grayling-Morphometrics-Public\Figure 3.tiff>" --prune-empty --tag-name-filter cat -- --all
WARNING: git-filter-branch has a glut of gotchas generating mangled history
         rewrites.  Hit Ctrl-C before proceeding to abort, then use an
         alternative filtering tool such as 'git filter-repo'
         (https://github.com/newren/git-filter-repo/) instead.  See the
         filter-branch manual page for more details; to squelch this warning,
         set FILTER_BRANCH_SQUELCH_WARNING=1.
Proceeding with filter-branch...

Rewrite 97aaa88a86263a6127b67f2eb03141d7b5394e58 (1/17) (0 seconds passed, remaining 0 predicted)    C:/Program Files/Git/mingw64/libexec/git-core\git-filter-branch: eval: line 439: syntax error near unexpected token `newline'
C:/Program Files/Git/mingw64/libexec/git-core\git-filter-branch: eval: line 439: `git rm --cached --ignore-unmatch <C:\Users\npwil\OneDrive\Desktop\School\Grad School\Side Quests\Grayling morphometrics project\Arctic-Grayling-Morphometrics-Public\Figure 3.tiff>'

C:\Users\npwil\OneDrive\Desktop\School\Grad School\Side Quests\Grayling morphometrics project\Arctic-Grayling-Morphometrics-Public>git filter-branch --force --index-filter "git rm --cached --ignore-unmatch <C:\Users\npwil\OneDrive\Desktop\School\Grad School\Side Quests\Grayling morphometrics project\Arctic-Grayling-Morphometrics-Public\Figure 3 option2.tiff>" --prune-empty --tag-name-filter cat -- --all
WARNING: git-filter-branch has a glut of gotchas generating mangled history
         rewrites.  Hit Ctrl-C before proceeding to abort, then use an
         alternative filtering tool such as 'git filter-repo'
         (https://github.com/newren/git-filter-repo/) instead.  See the
         filter-branch manual page for more details; to squelch this warning,
         set FILTER_BRANCH_SQUELCH_WARNING=1.
Proceeding with filter-branch...

Rewrite 97aaa88a86263a6127b67f2eb03141d7b5394e58 (1/17) (0 seconds passed, remaining 0 predicted)    C:/Program Files/Git/mingw64/libexec/git-core\git-filter-branch: eval: line 439: syntax error near unexpected token `newline'
C:/Program Files/Git/mingw64/libexec/git-core\git-filter-branch: eval: line 439: `git rm --cached --ignore-unmatch <C:\Users\npwil\OneDrive\Desktop\School\Grad School\Side Quests\Grayling morphometrics project\Arctic-Grayling-Morphometrics-Public\Figure 3 option2.tiff>'

C:\Users\npwil\OneDrive\Desktop\School\Grad School\Side Quests\Grayling morphometrics project\Arctic-Grayling-Morphometrics-Public>git push origin --force --all
Enumerating objects: 23, done.
Counting objects: 100% (21/21), done.
Delta compression using up to 8 threads
Compressing objects: 100% (15/15), done.
Writing objects: 100% (15/15), 2.83 MiB | 520.00 KiB/s, done.
Total 15 (delta 10), reused 0 (delta 0), pack-reused 0
remote: Resolving deltas: 100% (10/10), completed with 4 local objects.
remote: error: Trace: cad7233c80ce0198e8fa064131f52e06195e1e736b27898a53d05d76f23dbbfb
remote: error: See https://gh.io/lfs for more information.
remote: error: File Figure 3 option2.tiff is 314.71 MB; this exceeds GitHub's file size limit of 100.00 MB
remote: error: File Figure 3.tiff is 314.71 MB; this exceeds GitHub's file size limit of 100.00 MB
remote: error: GH001: Large files detected. You may want to try Git Large File Storage - https://git-lfs.github.com.
To https://github.com/WillSamuel18/Arctic-Grayling-Morphometrics-Public
 ! [remote rejected] main -> main (pre-receive hook declined)
error: failed to push some refs to 'https://github.com/WillSamuel18/Arctic-Grayling-Morphometrics-Public'

C:\Users\npwil\OneDrive\Desktop\School\Grad School\Side Quests\Grayling morphometrics project\Arctic-Grayling-Morphometrics-Public>
C:\Users\npwil\OneDrive\Desktop\School\Grad School\Side Quests\Grayling morphometrics project\Arctic-Grayling-Morphometrics-Public>

**Comment by Muubii** [comment]
My friend had that same problem what you can do you select a few files and then push it

**Comment by nagasriramnani** [comment]
To resolve the issue of large files that are preventing you from pushing to GitHub, you'll need to remove the files from your git history. Since you've already deleted the large files from your local folder, you need to make sure these deletions are reflected in your git history.

Here's how you can remove the large files from your git history and push your current work:

Remove Large Files from History:

Use the git filter-branch command or the BFG Repo-Cleaner, which is a simpler and faster tool. The BFG Repo-Cleaner specifically is a tool that's designed for cleaning up large files or sensitive data from git repositories.

If you choose to use git filter-branch, the command would be something like:

sh

git filter-branch --force --index-filter \
"git rm --cached --ignore-unmatch path_to_your_large_file.tiff" \
--prune-empty --tag-name-filter cat -- --all
Replace path_to_your_large_file.tiff with the path to the large file you want to remove. Note that this operation will rewrite your git history, which can be disruptive to your collaborators if you're working with others.

Force Push the Cleaned Repository:

After cleaning your repository, you will need to force push your changes to overwrite the history on GitHub.

sh

git push origin main --force
Use Git LFS Moving Forward:

Since you have files that exceed GitHub's file size limit, consider using Git Large File Storage (LFS) for handling large files. You can track large files locally with Git LFS and push them to GitHub without encountering size limitations.

sh
git lfs install
git lfs track "*.tiff"
git add .gitattributes
Commit the .gitattributes file which now contains your Git LFS tracking rules, and then you can commit and push as you normally would.

Remember that modifying git history, especially with force pushes, can cause issues for any collaborators working on the same repository. Make sure to communicate with your team about these changes, and ensure everyone has the correct and updated history after you've made your changes.

Since you're using RStudio, after cleaning your git history via the command line, you can then return to RStudio and continue your normal git workflow from there.
